File: crates/convertor/src/url/url_result.rs
use crate::url::convertor_url::ConvertorUrl;
use serde::{Deserialize, Serialize};
use std::fmt;
use std::fmt::Display;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UrlResult {
    pub raw_url: ConvertorUrl,
    pub raw_profile_url: ConvertorUrl,
    pub profile_url: ConvertorUrl,
    pub sub_logs_url: ConvertorUrl,
    pub rule_providers_url: Vec<ConvertorUrl>,
}

impl Display for UrlResult {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "{}", self.raw_url.desc)?;
        writeln!(f, "{}", self.raw_url)?;
        writeln!(f, "{}", self.profile_url.desc)?;
        writeln!(f, "{}", self.profile_url)?;
        writeln!(f, "{}", self.raw_profile_url.desc)?;
        writeln!(f, "{}", self.raw_profile_url)?;
        writeln!(f, "{}", self.sub_logs_url.desc)?;
        writeln!(f, "{}", self.sub_logs_url)?;
        for url in &self.rule_providers_url {
            writeln!(f, "{}", url.desc)?;
            writeln!(f, "{url}")?;
        }
        Ok(())
    }
}

File: dashboard/src/app/common/model/url_result.ts
import Cloneable from "../base/cloneable";
import Equatable from "../base/equals";
import Serializable from "../base/serializable";
import { ConvertorUrl } from "./convertor_url";

export class UrlResult implements Cloneable<UrlResult>, Equatable<UrlResult>, Serializable {
    public constructor(
        public raw_url: ConvertorUrl,
        public raw_profile_url: ConvertorUrl,
        public profile_url: ConvertorUrl,
        public rule_providers_url: ConvertorUrl[],
    ) {
    }

    public static deserialize(result: UrlResult) {
        return new UrlResult(
            ConvertorUrl.deserialize(result.raw_url),
            ConvertorUrl.deserialize(result.raw_profile_url),
            ConvertorUrl.deserialize(result.profile_url),
            result.rule_providers_url.map(ConvertorUrl.deserialize),
        );
    }

    public clone(): UrlResult {
        return new UrlResult(
            this.raw_url.clone(),
            this.raw_profile_url.clone(),
            this.profile_url.clone(),
            this.rule_providers_url.map((rp) => rp.clone()),
        );
    }

    public equals(other?: UrlResult): boolean {
        if (!other) return false;
        return this.raw_url.equals(other.raw_url)
            && this.raw_profile_url.equals(other.raw_profile_url)
            && this.profile_url.equals(other.profile_url)
            && this.rule_providers_url.length === other.rule_providers_url.length
            && this.rule_providers_url.every((rp, index) => rp.equals(other.rule_providers_url[index]));

    }

    public serialize(): any {
        return {
            raw_url: this.raw_url.serialize(),
            raw_profile_url: this.raw_profile_url.serialize(),
            profile_url: this.profile_url.serialize(),
            rule_providers_url: this.rule_providers_url.map((rp) => rp.serialize()),
        };
    }

}
